lib/CL/devices/cuda/pocl-cuda.c

```diff
@@ -216,6 +218,226 @@ pocl_cuda_error (CUresult result, unsigned line, const char *func,
+      || parsed_num == LLONG_MAX || parsed_num == LLONG_MIN)
+    return 0;
+
+  // TODO: Check the value is a known SM version?
+  return parsed_num;
+}
+
+struct cuda_ptx_version_pair_s
+{
+  unsigned cuda_version;
+  unsigned ptx_version;
+};
+
+// Maps CUDA driver version to PTX version the driver supports at most
+// (there doesn't seem to be a way to query supported PTX versions).
+const struct cuda_ptx_version_pair_s cuda_ptx_version_map[] = {
+  // Based on [PTX ISA v8.5 - 12. Release Notes]. For simplicity,
+  // minor CUDA versions are (mostly) skipped for now.
+  { 12000, 80 }, { 11000, 70 }, { 10000, 63 }, { 9000, 60 }, { 8000, 50 },
+  { 7000, 42 },  { 6000, 40 },  { 5000, 31 },  { 4010, 30 }, { 4000, 23 },
+  { 3000, 20 },  { 2000, 12 },  { 1000, 10 },  { 0, 0 } // Sentinel.
+};
+
+// Return a PTX version range that is supported by the CUDA driver for the
+// specified SM architecture.
+static void
+get_valid_ptx_version_range (unsigned cuda_version, unsigned sm_version,
+                             unsigned *min_ptx_out, unsigned *max_ptx_out)
+{
+  assert (min_ptx_out && max_ptx_out);
+
+  // SM version sets lower bound. PTX version needs to be high enough
+  // target the given SM version.
+  unsigned min_ptx = 0;
+
+  // Based on [PTX ISA v8.5 - 12. Release Notes].
+  if (sm_version >= 89)
+    min_ptx = 78;
+  else if (sm_version >= 87)
+    min_ptx = 74;
+  else if (sm_version >= 86)
+    min_ptx = 71;
+  else if (sm_version >= 80)
+    min_ptx = 70;
+  else if (sm_version >= 75)
+    min_ptx = 63;
+  else if (sm_version >= 72)
+    min_ptx = 61;
+  else if (sm_version >= 70)
+    min_ptx = 60;
+  else if (sm_version >= 60)
+    min_ptx = 50;
+  else if (sm_version >= 53)
+    min_ptx = 42;
+  else if (sm_version >= 52)
+    min_ptx = 41;
+  else if (sm_version >= 50)
+    min_ptx = 40;
+  else if (sm_version >= 35)
+    min_ptx = 31;
+  else if (sm_version >= 30)
+    min_ptx = 30;
+  else if (sm_version >= 20)
+    min_ptx = 20;
+  else if (sm_version >= 13)
+    min_ptx = 12;
+  else if (sm_version >= 11)
+    min_ptx = 10;
+
+  // CUDA runtime sets upper bound. PTX version may not be higher than
+  // what the runtime recognizes.
+
+  unsigned max_ptx = 0;
+  for (unsigned i = 0; cuda_ptx_version_map[i].cuda_version; i++)
+    if (cuda_version >= cuda_ptx_version_map[i].cuda_version)
+      {
+        max_ptx = cuda_ptx_version_map[i].ptx_version;
+        break;
+      }
+
+  assert (min_ptx && max_ptx
+          && "Failed to determine complete PTX version bounds!");
+  assert (min_ptx <= max_ptx && "Determined contradicting PTX version range!");
+
+  *min_ptx_out = min_ptx;
+  *max_ptx_out = max_ptx;
+}
+
+static void
+choose_sm_ptx_version_combo (unsigned cuda_version, cl_device_id dev,
+                             pocl_cuda_device_data_t *data)
+{
+  // Choose SM version to target.
+
+  unsigned selected_sm = 0;
+
+  const char *user_sm = pocl_get_string_option ("POCL_CUDA_GPU_ARCH", NULL);
```

Comment:
This should be `POCL_CUDA%d_GPU_ARCH` where %d is the index of the cuda device. I don't know if we support multiple CUDA devices currently, but each of them could be different. This can be thus postponed.

Reply:
I'd leave it out of this PR. Note, this isn't a new feature.

Support for multiple CUDA devices is claimed in the docs:
```
Use POCL_DEVICES=CUDA to select only CUDA devices. If the system has 
more than one GPU, specify the CUDA device multiple times 
(e.g. POCL_DEVICES=CUDA,CUDA for two GPUs).
```

`POCL_CUDA%d_GPU_ARCH` approach doesn't seem to scale well. Perhaps we could specify different architectures in the form of `POCL_CUDA_GPU_ARCH=sm_XX,sm_YY,...`.